crates/astria-composer/src/searcher/bundler.rs

```diff
@@ -0,0 +1,135 @@
+    fn push(&mut self, seq_action: SequenceAction) -> Result<(), SizedBundleError> {
+        let seq_action_size = seq_action.data.len() + ROLLUP_ID_LEN;
+        if self.curr_size + seq_action_size > self.max_size {
+            return Err(SizedBundleError::NotEnoughSpace(seq_action));
+        }
+        self.buffer.push(Action::Sequence(seq_action));
+        self.curr_size += seq_action_size;
+        Ok(())
+    }
+
+    /// Replace self with a new empty bundle, returning the old bundle.
+    fn flush(&mut self) -> SizedBundle {
+        mem::replace(self, Self::new(self.max_size))
+    }
+
+    /// Consume self and return the underlying buffer of actions.
+    pub(super) fn into_actions(self) -> Vec<Action> {
+        self.buffer
+    }
+
+    pub(super) fn is_empty(&self) -> bool {
+        self.buffer.is_empty()
+    }
+
+    pub(super) fn len(&self) -> usize {
+        self.buffer.len()
+    }
+}
+
+/// Manages the bundling of sequence actions into `SizedBundle`s. A `SizedBundle` is flushed and
+/// added to the `finished` queue when an incoming `SequenceAction` won't fit in the current bundle.
+/// The `finished` queue operates in FIFO order, where `SizedBundle`s are added to the back and
+/// taken off from the front.
+pub(super) struct Bundler {
+    /// The current bundle being built.
+    pub(super) curr_bundle: SizedBundle,
+    /// The queue of bundles that have been built but not yet sent to the sequencer.
+    pub(super) finished: VecDeque<SizedBundle>,
+}
+
+impl Bundler {
+    pub(super) fn new(max_size: usize) -> Self {
+        Self {
+            curr_bundle: SizedBundle::new(max_size),
+            finished: VecDeque::new(),
+        }
+    }
+
+    /// Buffer `seq_action` into the current bundle. If the bundle won't fit `seq_action`, flush
+    /// `curr_bundle` into the `finished` queue and start a new bundle
+    pub(super) fn push(&mut self, seq_action: SequenceAction) {
+        let seq_action_size = seq_action.data.len() + ROLLUP_ID_LEN;
+        if let Err(SizedBundleError::NotEnoughSpace(seq_action)) = self.curr_bundle.push(seq_action)
+        {
+            // if the bundle is full, flush it and start a new one
+            self.finished.push_back(self.curr_bundle.flush());
+            // TODO: this will error if the seq_action size is larger than the max bundle size -
+            // should panic if its reached this point
+            self.curr_bundle
+                .push(seq_action)
+                .expect("seq_action should not be larger than max bundle size");
+        }
+        debug!(
+            new_bundle_size = ?self.curr_bundle.curr_size,
+            seq_action_size = ?seq_action_size,
+            "bundled new sequence action"
+        );
+    }
+
+    /// Get the next bundle from the `finished` queue. If the queue is empty, flush `curr_bundle`
+    pub(super) fn get_next(&mut self) -> Ready<Option<SizedBundle>> {
+        future::ready(self.finished.pop_front())
+    }
```

Comment:
the fact that this always returns `Ready` is sus to me. is this going to break the select statement @SuperFluffy ?

you should probably implement `Future` for `Bundler` or some other type inside `Bundler`, eg
```
impl Future for Bundler {
    type Output = SizedBundle;

   fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
        match self.unpin().finished.pop_front() {
             Some(bundle) => Poll::Ready(bundle),
             None => Poll::Pending,
        }
}
```

maybe there's a type for this already in the futures lib somewhere? will look

Reply:
This was @SuperFluffy 's suggestions so let's see what he says after reviewing the whole PR

Reply:
I am not a fan of the naming here.

`get_*` taking a mutable borrow is very surprising. It should also not return a ready-future.

IMO this would be a better signature:
```rust
fn pop(&mut self) -> Option<Vec<Action>>
```

Also note the return value: IMO `SizedBundle` should remain an implementation detail. It's fine if downstream users can rely on receiving a bundle of the correct size.